### DISCUSSION TITLE
GRPC High Latency Issue

### DISCUSSION THREAD

**Discussion by dnjp** [linkerd/linkerd2#4671] [General] [answered]
![latency](https://user-images.githubusercontent.com/37786372/85793510-1e56a080-b703-11ea-99d3-c8ca1d26d514.png)

I am trying to determine why the latency for this simple GRPC service running in Kubernetes with Linkerd is so high. I've prepared a minimal example of the service that was used for these latency tests which you can see in [this gist](https://gist.github.com/danieljamespost/579843b1a7ea2ef6a14ab96c7db6dab0). As you can see in the gist, the client ("augmentor" in the image above) is receiving an HTTP request, sending an Echo message to the server ("bidder" in the image above) and the server simply returns the message it received. Under less than 100rps, the P99 latency is around 30ms. If I increase to about 1,200rps the latency skyrockets and will get up to around 200ms.

I have conducted this test on a fresh GKE Cluster running only these two services, deployed with the configuration you will find in the gist. I've installed Linkerd by following the quickstart guide and the output of `linkerd check` is below. When viewing logs with `linkerd tap` the requests are coming through properly as HTTP/2 with TLS enabled. I've also verified this is the case by using Wireshark to view the contents of these requests. If I increase rps further and have tracing enabled (using Jaeger), the traces look like this in the worst case: 
![screenshot_000](https://user-images.githubusercontent.com/37786372/85795204-eb61dc00-b705-11ea-8493-fdf96e6da6fa.png)

Given how simple this example application is, I'm inclined to think the issue is likely networking related. Is there something I have not configured properly that would cause this kind of behavior?

```
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ tap api service is running

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
√ control plane is up-to-date
√ control plane and cli versions match

linkerd-addons
--------------
√ 'linkerd-config-addons' config map exists

linkerd-grafana
---------------
√ grafana add-on service account exists
√ grafana add-on config map exists
√ grafana pod is running

Status check results are √
```

**Comment by ihcsim** [comment]
200ms does sound a bit slow. The first thing that stood out in your YAML is that you assigned about 4 CPUs to your client and server pods. Did you configure the Linkerd proxy with CPU requests? Make sure your proxies aren't starving. You can use either the `linkerd inject` or [proxy configuration annotations](https://linkerd.io/2/reference/proxy-configuration/) to set these values. It will be interesting to see each container's resource usage.

Also, what does the latency distribution look like without Linkerd? If you haven't already, might want to run the test with a continuous traffic flow for a longer period of time (e.g., 1000 rps for > 5 mins), to get a better sense of the distribution trend.

**Reply by dnjp** [reply]
Yes, I've tried increasing these settings with the following annotations and received similar results.
```
      annotations:
        config.linkerd.io/proxy-cpu-limit: "1.5"
        config.linkerd.io/proxy-cpu-request: "1"
        config.linkerd.io/proxy-memory-limit: 2Gi
        config.linkerd.io/proxy-memory-request: 1Gi
        linkerd.io/inject: enabled
```
We have a hard latency requirement for this request to take at most 10ms at around 2-3k rps per pod. Do you have any recommendations for debugging what could be taking so long with the current configuration? From the trace I shared it looks like it is hanging at the networking layer. I can try testing again without Linkerd to get some default latency measurements.

**Reply by dnjp** [reply]
Securing GRPC with TLS got latency down to around 20ms and removing the VPC got latency down to sub 10ms. Here is how this is currently configured:

Client:
```
	creds, err := credentials.NewClientTLSFromFile("certs/server.pem", "")
	if err != nil {
		log.Fatalf("Failed to setup TLS: %+v", err)
	}
```

Server:
```
	creds, err := credentials.NewServerTLSFromFile("certs/server.pem", "certs/server.key")
	srv := grpc.NewServer(grpc.Creds(creds))
	if err != nil {
		log.Fatalf("Failed to setup TLS: %v", err)
	}
```

Using `WithInsecure()` drastically reduced performance. Is there a more elegant way to handle this using Linkerd?

**Reply by dnjp** [reply]
It's also worth noting that handling TLS in the application appears to prevent any metrics from being recorded by Linkerd. Is there a way around that?

**Reply by ihcsim** [reply]
> We have a hard latency requirement for this request to take at most 10ms at around 2-3k rps per pod

2-3k rps (inbound + outbound) shouldn't be a problem. At the moment, the proxy is still single threaded. There are some upcoming work to improve this.

> Securing GRPC with TLS got latency down to around 20ms and removing the VPC got latency down to sub 10ms.

I am surprised that `WithInsecure()` would degrade the performance. If anything, I'd suspect it's `WithBlock()`. Since these are unary calls, your client will create new connections and get blocked [here](https://gist.github.com/danieljamespost/579843b1a7ea2ef6a14ab96c7db6dab0#file-client-go-L51), right? I wonder if this will affect the overall latency distribution.

Also, can you share the latency distribution of your workloads as seen on the Linkerd Grafana dashboard? The `Linkerd Deployment` dashboard should have something relevant. You can filter the charts by namespaces. Any idea what the baseline latency (without Linkerd) looks like?

I will also be interested in the pods/containers resource utilization. `kubectl top po` is a good start, and if you are on GKE, Stackdriver should provide those metrics OOB.

And yes, once you TLS'd your services, the traffic will be encrypted. Linkerd can't MITM it, hence, it won't pick up any metrics.

**Reply by ihcsim** [reply]
Are you able to share your container image with us, for repro purposes?

**Reply by dnjp** [reply]
The documentation for this: "WithBlock returns a DialOption which makes caller of Dial blocks until the underlying connection is up. Without this, Dial returns immediately and connecting the server happens in background." So it is simply blocking until the connection is acquired, not making each call with the client blocking. Just to be sure, I tried this test with the option disabled and it did not have any affect on the latency.

I tried using `WithInsecure`and relying on mTLS with Linkerd using the default VPC in GCP and I'm still getting 150ms+ latency. Is there something that I would have configured incorrectly that would cause that kind of latency with mTLS? I'd greatly prefer to be able to use mTLS with Linkerd and get metrics about each pod.

**Reply by ihcsim** [reply]
Nope, there isn't much to configure. We just need to figure out the source of the latency. Earlier I had our tracing expert looked at your tracing screenshot, who pointed out that the client's outbound proxy waited for > 1.5s before sending the request, per the `augmentor` and first `linkerd-proxy` spans. We would like to find out what the proxy is doing. Can you share the `augmentor` proxy debug logs, and the result of `linkerd metrics`? You can follow these [instructions](https://linkerd.io/2/tasks/modifying-proxy-log-level/) to modify the proxy log level. Thanks!

**Reply by dnjp** [reply]
Unfortunately I'm not at liberty to share the dockerfiles for these applications, but they are based on the golang:1.14-alpine image and use [protoc-gen-go](https://github.com/golang/protobuf/tree/master/protoc-gen-go).

As for resource utilization, the deployment is far from resource constrained, they are using 636m CPU and 44Mi memory with the limits set in the gist.

**Reply by dnjp** [reply]
> Can you share the `augmentor` proxy debug logs, and the result of `linkerd metrics`? You can follow these [instructions](https://linkerd.io/2/tasks/modifying-proxy-log-level/) to modify the proxy log level. Thanks!

Not a problem. I needed to change the actual endpoint being requested, but these are the logs I'm seeing from the linkerd-proxy sidecar:

[client-logs.txt](https://github.com/linkerd/linkerd2/files/4839644/augmentor-logs.txt)
[server-logs.txt](https://github.com/linkerd/linkerd2/files/4839645/bidder-logs.txt)

**Reply by dnjp** [reply]
Here are the metrics collected:

[client-metrics.txt](https://github.com/linkerd/linkerd2/files/4839661/client-metrics.txt)
[server-metrics.txt](https://github.com/linkerd/linkerd2/files/4839662/server-metrics.txt)

**Reply by dnjp** [reply]
Here are the logs and metrics with the `WithInsecure` flag enabled using mTLS

[client-logs-mtls.txt](https://github.com/linkerd/linkerd2/files/4839688/augmentor-logs-mtls.txt)
[server-logs-mtls.txt](https://github.com/linkerd/linkerd2/files/4839689/bidder-logs-mtls.txt)
[client-metrics-mtls.txt](https://github.com/linkerd/linkerd2/files/4839690/client-metrics-mtls.txt)
[server-metrics.mtls.txt](https://github.com/linkerd/linkerd2/files/4839691/server-metrics.mtls.txt)

**Reply by dnjp** [reply]
Just following up - we are currently unable to fully utilize Linkerd due to the latency involved with using mTLS for gRPC connections. Were you able to find anything useful in the metrics/logs that would lead to an mTLS based solution so we can use Linkerd for this service?

**Reply by ihcsim** [reply]
Can you tell me more about your GKE setup? We will try to reproduce this.

1. Version of GKE
1. Node OS: ubuntu, cos etc.
1. Number of nodes. 
1. Are the nodes on the same node pool?
1. Do you have network policies?
1. You mentioned about VPC. So I assume it's the pods are using alias IP
1. Is this a private GKE clusters?
1. Do you have any customization on your `kube-dns` deployment?
1. Is this latency reproducible on another GKE cluster?

**Reply by dnjp** [reply]
>     1. Version of GKE
1.16.9-gke.6 

>     2. Node OS: ubuntu, cos etc.
cos
 
>     3. Number of nodes.
10 nodes
 
>     4. Are the nodes on the same node pool?
Yes
 
>     5. Do you have network policies?
No

>     6. You mentioned about VPC. So I assume it's the pods are using alias IP
That's correct

>     7. Is this a private GKE clusters?
No

>     8. Do you have any customization on your `kube-dns` deployment?
No

>     9. Is this latency reproducible on another GKE cluster?
Yes, we have an identical cluster in us-west1 that has these same issues. I also created a separate cluster for testing which had these same problems.

**Comment by ihcsim** [comment, marked as answer]
@danieljamespost So far, I haven't been able to reproduce the latency you experienced. At 1200 rps, the p99 latency hovers around 6ms to 9ms, with 32 concurrent connections:
![strest-grpc-latency-1200rps](https://user-images.githubusercontent.com/1330522/86184426-0dce6d80-bae9-11ea-8dcb-c0c6465c5403.png)

Here's the p99 trend over a period of 15 minutes:
![strest-grpc-latency-1200rps-grafana](https://user-images.githubusercontent.com/1330522/86184473-250d5b00-bae9-11ea-8dc1-76b29c14d3cb.png)

The CPU utilization is around 70% to 80% (with a limit of 1 CPU):
![strest-grpc-latency-1200rps-cpu](https://user-images.githubusercontent.com/1330522/86184541-48380a80-bae9-11ea-95dd-98171fa6ac4d.png)

While the memory usage is near negligible (with a limit of 2GB):
![strest-grpc-latency-1200-rps-memory](https://user-images.githubusercontent.com/1330522/86184550-4cfcbe80-bae9-11ea-972b-c61ac9b868f5.png)

I tested Linkerd 2.8.1 on a GKE cluster comprised of [`n1-standard-1`](https://cloud.google.com/compute/docs/machine-types) nodes, using the following resource requirements on the client and server proxies:

CPU: 
  * limit: 1
  * request: 0.25
Memory
  * limit: 2GB
  * request: 1GB

Also, nothing stood out to me in the attached logs and metrics so far.

Are you able to gather the CPU usage of the proxy? I wonder if the latency you saw was caused by CPU spikes on the proxy.

**Reply by dnjp** [reply]
Thank you for taking the time to try to reproduce the issue. I'm assuming you tested this out in `us-east1` using the default VPC provisioned when you create the cluster? Let me know if that isn't the case. Also, did you use the `WithInsecure` option or did you use TLS certs?

I can run another test early next week to get some more data around the proxy.

**Reply by ihcsim** [reply]
Would you mind trying out the latest edge proxy? There are some improvements in there that might be relevant. Keep your current Linkerd control plane setup as-is, and just annotate your client and server deployments (in the **pod template spec**) with `config.linkerd.io/proxy-version: edge-20.7.1`. The pods will be restarted with the latest proxy image. Give that a try and let us know how it goes.

Yes, I used the default GKE VPC setup in my test. But my cluster is in the `us-west1` region. The tool that I used is called [strest-grpc](https://github.com/buoyantio/strest-grpc), which also uses the `WithInsecure` option on the [client side](https://github.com/BuoyantIO/strest-grpc/blob/3af6d246dc348ab063608898d979622088748ead/client/client.go#L854).

**Comment by dnjp** [comment]
@ihcsim I apologize for taking so long to get back to this issue. We had to hold off on deploying the new GRPC service that I was going to test with. However, I'm now running edge-20.8.4 and the latency issues have gone away and we are getting excellent performance. There are other content-type issues I'm looking into, but the performance is great. Thank you for your help.
